typescript/schedules.md

```diff
@@ -0,0 +1,1255 @@
+
+const { schedules, nextPageToken } = await client.listByPage({
+  pageSize: 50,
+  nextPageToken: 'base64',
+});
+
+for await (const schedule: Schedule of client.list()) {
+  const { id, memo, searchAttributes } = schedule;
+  // ...
+}
+```
+
+### Types
+
+```ts
+import { DataConverter, LoadedDataConverter } from '@temporalio/common';
+import { loadDataConverter } from '@temporalio/internal-non-workflow-common';
+import {
+  composeInterceptors,
+  Headers,
+  Replace,
+  RetryPolicy,
+  SearchAttributes,
+  Workflow,
+} from '@temporalio/internal-workflow-common';
+import { temporal } from '@temporalio/proto';
+import os from 'os';
+import { Connection } from './connection';
+import { ScheduleClientCallsInterceptor, ScheduleClientInterceptors } from './interceptors';
+import { ConnectionLike, Metadata, WorkflowService } from './types';
+import { WorkflowHandle, WorkflowStartOptions } from './workflow-client';
+
+import type { NonNegativeInteger, RequireAtLeastOne } from 'type-fest';
+
+// TODO is a non-generic NonNegativeInteger possible? The ones below error due to no type argument
+export type PositiveInteger = Exclude<NonNegativeInteger, 0>;
+
+/**
+ * @format number of milliseconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
+ */
+export type Milliseconds = string | NonNegativeInteger;
+
+/**
+ * @format number of seconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
+ */
+export type Seconds = string | NonNegativeInteger;
+
+export interface UpdateScheduleOptions {
+  /**
+   * @default TODO
+   */
+  retry?: RetryPolicy;
+}
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  startAt: Date;
+  endAt: Date;
+
+  /** Override Overlap Policy for this request. */
+  overlapPolicyOverride?: ScheduleOverlapPolicy;
+}
+
+/**
+ * Handle to a single Schedule
+ */
+export interface ScheduleHandle {
+  /**
+   * This Schedule's identifier
+   */
+  readonly id: string;
+
+  /**
+   * Update the Schedule
+   */
+  update(mutateFn: (schedule: ScheduleDescription) => void | false, options?: UpdateScheduleOptions): Promise<void>;
+
+  /**
+   * Delete the Schedule
+   */
+  delete(): Promise<void>;
+
+  /**
+   * Trigger an Action to be taken immediately
+   */
+  trigger(): Promise<void>;
+
+  /**
+   * Run though the specified time period(s) and take Actions as if that time passed by right now, all at once. The
+   * Overlap Policy can be overridden for the scope of the Backfill.
+   */
+  backfill(options: Backfill | Backfill[]): Promise<void>;
+
+  /**
+   * Pause the Schedule
+   */
+  pause(note?: string): Promise<void>;
+
+  /**
+   * Unpause the Schedule
+   */
+  unpause(note?: string): Promise<void>;
+
+  /**
+   * Fetch the Schedule's description from the Server
+   */
+  describe(): Promise<ScheduleDescription>;
+
+  /**
+   * Get a handle to the most recent Action started
+   */
+  lastAction<T extends ScheduleActionType>(): Promise<HandleFor<T>>;
```

Comment:
Can you clarify the API calls this makes and what this does? Just here in comments is fine, I'm trying to confirm the value of this.

Reply:
I see. I would prefer we avoid shortcuts that actually make service calls. If you want to add a `lastAction` helper on the description response, that could make more sense than a shortcut here that makes a whole separate network call.

Reply:
I would remove this helper for now

Reply:
also keep in mind that the runId in the describe result is the first-execution-runid, not necessary the current runid. so that handle might not be what you want.

Reply:
Thanks for the reminder David, I'm renaming runId -> firstExecutionRunId.

Why remove helper from description? It still seems like a common use case and improved DX:

```ts
description = await handle.describe()
handle = description.lastAction()
// or: 
handle = description.lastAction({ mostRecentInChain: false })
```

```ts
description = await handle.describe()
workflowClient = new WorkflowClient()
recent = description.info.recentActions
last = recent[recent.length - 1]
we = last.workflow!
handle = workflowClient.getHandle(we.worfklowId, null, { firstExecutionRunId: we.firstExecutionRunId })
// or:
handle = workflowClient.getHandle(we.worfklowId, we.firstExecutionRunId)
```

Reply:
The reason for me would be that you'd have to instantiate a WorkflowClient implicitly.
The user would probably want to control that instantiation.
We wouldn't have this problem if we went with a single client but that ship has sailed IMO.

Reply:
I'm fine with the general idea though, some users might find this helpful.

Reply:
what about a `description.lastAction` getter that just returns the last of recentActions? (doesn't create a handle)